sql/sql_connect.cc

```diff
@@ -255,6 +262,105 @@ void time_out_user_resource_limits(THD *thd, USER_CONN *uc)
   DBUG_VOID_RETURN;
 }
 
+static uint64 get_connection_delay_time(uint32 count)
+{
+  /*
+   * Based on that invalid password errors will cause user can't connect server
+  ever and server close connection if it got message from client after
+  wait_timeout
```

Comment:
as you've found out, this is not correct, the delay can be longer than wait_timeout and the server will not disconnect. We artificially limit the delay by wait_timeout. So, better reflect that in the comment, for example
```c
/*
  connect_timeout is designed to limit the connect time, every single network read and
  write is limited by it. Thus we'll say here that the maximal delay is limited by the
  connect_timeout too, and if it takes longer - the connection times out and the user
  cannot connect at all. This means that the delay must reach the connect_timeout
  when the count reaches max_password_errors
*/